### DISCUSSION TITLE
Is it possible to show more meaningful tab titles in the ShowTabNavigatior view?

### DISCUSSION THREAD

**Discussion by joshuahicks** [wezterm/wezterm#4673] [Q&A] [answered]
### What Operating System(s) are you running on?

macOS

### WezTerm version

wezterm 20230712-072601-f4abf8fd

### Ask your question!

Hi,

I use [ShowTabNaviagtor](https://wezfurlong.org/wezterm/config/lua/keyassignment/ShowTabNavigator.html) to navigate between my tabs. If I have multiple vim tabs open, it's difficult to know which tab is what, since they all display the name of the application running on that tab. 

Is it possible to change what is displayed for the tabs in this view? It would be helpful if I could see the folder, instead, for example.

![Screen Shot 2023-12-13 at 3 59 46 PM](https://github.com/wez/wezterm/assets/13204883/3ed00960-e7ab-4efc-8468-193d1b9f793b)

I see information on [set_title()](https://wezfurlong.org/wezterm/config/lua/MuxTab/set_title.html). I'm not sure how to use this, but I will play around with this, while I wait for a potential answer. Thank you

Note: [Here](https://github.com/joshuahicks/.dotfiles/blob/main/.config/wezterm/wezterm.lua) is my wezterm config for reference

**Comment by wez** [comment, marked as answer]
I use a shell pre-exec or pre-command hook to emit OSC 0 to set the title. That way the shell will update the terminal title just before it launches any program.

https://wezfurlong.org/wezterm/escape-sequences.html#operating-system-command-sequences shows a raw example of using OSC 0.

Here's an excerpt from my dotfiles:

```bash
# Removes ANSI escape sequences from a string
function _dotfiles_ansi_dequote()
{
  # strip-ansi-escapes is provided in rpm and deb installs of wezterm
  if hash strip-ansi-escapes 2>/dev/null ; then
    echo "$1" | strip-ansi-escapes
  else
    echo "$1" | perl -pe "s/\\x1b\[[0-9]+(;[0-9]+)*[km]//g"
  fi
}

# This function was originally born in the '90s and deals
# with terminals that are no longer common.
# You can likely prune this to just the xterm and tmux
# portions.
function _dotfiles_set_window_title()
{
  local title
  title=`_dotfiles_ansi_dequote "$1" | perl -pe "s/;//g"`
  case $TERM in
    xterm*|rxvt*|Eterm|eterm|aixterm|dtterm)
      printf "\033]0;%s\a" "$title"
      ;;
    iris-ansi)
      printf "\033P1.y%s\033\\" "$title"
      ;;
    sun-cmd)
      printf "\033]l%s\033\\" "$title"
      ;;
    hpterm)
      printf "\033&f0k%dD%s" "${#title}" "${title}"
      ;;
    screen*|tmux*)
      # screen and tmux window title
      printf "\033k%s\033\\" "$title"
      # set the tmux pane title too
      [[ -n "$TMUX" ]] && printf "\033]2;%s\033\\" "$title"
      ;;
  esac
}

# This is how it is enabled for zsh.
# For bash, I use https://github.com/rcaloras/bash-preexec to enable the
# same functionality in the same way
function precmd()
{
  # Formats the title with useful context
  _dotfiles_set_window_title "${USER}@${HOSTNAME%%.*}:${PWD/$HOME/~}"
}
```

**Reply by joshuahicks** [reply]
Awesome, thanks @wez! I'll give this a shot

**Reply by miguno** [reply]
This code snippet also worked for me when I was troubleshooting an issue where the window name and the "tab title" within wezterm wouldn't correctly update when navigating between directories.  The name/title would be stuck to whatever was set by neovim (e.g., keep showing `COMMIT_EDITMSG`).  The root cause could have been a broken `.zshrc` configuration or whatever, but anyways adding the code above to `.zshrc` (to explicitly set the name/title exactly how I want it) fixed it.

However, I updated the code to rely on `sed` instead of `perl` to remove ANSI escape sequences from a string:

```shell
# Removes ANSI escape sequences from a string
function _dotfiles_ansi_dequote()
{
  # strip-ansi-escapes is provided in rpm and deb installs of wezterm
  if hash strip-ansi-escapes 2>/dev/null ; then
    echo "$1" | strip-ansi-escapes
  else
    # CHANGE HERE (sed instead of perl)
    echo "$1" | sed -E "s/\\x1b\[[0-9]+(;[0-9]+)*[km]//g"
  fi
}
```

Thank you!
